### PR TITLE
Add beatmap header to leaderboard response

### PR DESCRIPTION
**Pull request by ascenttree**
Starts a basic stub for leaderboard response (packet id 25), also adding a reader object for the leaderboard request (also id 25)

See #25

### CODE REVIEW

File: hnet/handler.go [hunk 1]
```diff
@@ -265,6 +265,33 @@ func handleUserRelationshipRemove(stream *common.IOStream, player *Player) error
 	return nil
 }
 
+func handleLeaderboardRequest(stream *common.IOStream, player *Player) error {
+	request := ReadLeaderboardRequest(stream)
```
Comment (Lekuruu) [inline hnet/handler.go:269]: Don't forget to check if the request is `nil`

File: hnet/handler.go [hunk 2]
```diff
@@ -265,6 +265,33 @@ func handleUserRelationshipRemove(stream *common.IOStream, player *Player) error
...
+	response := &LeaderboardResponse{
+		BeatmapChecksum: request.BeatmapChecksum,
+		Unknown:         request.Unknown,
+		NeedsUpdate:     false,
+		Status:          common.StatusNotSubmitted,
+		ShowScores:      request.ShowScores,
+	}
+
+	beatmap, err := common.FetchBeatmapById(int(request.BeatmapId), player.Server.State)
+	if err != nil {
```
Comment (Lekuruu) [inline hnet/handler.go:281]: You don't return after checking for `err != nil`. Not sure if this is intended?

**Review by Lekuruu — CHANGES_REQUESTED** [review]
A few things I saw on first glance, will have to do a proper review once I am home.

File: hnet/handler.go [hunk 3]
```diff
@@ -265,6 +265,39 @@ func handleUserRelationshipRemove(stream *common.IOStream, player *Player) error
 	return nil
 }
 
+func handleLeaderboardRequest(stream *common.IOStream, player *Player) error {
+	request := ReadLeaderboardRequest(stream)
+
+	if request == nil {
+		return fmt.Errorf("failed to read leaderboard request")
+	}
+
```
Comment (Lekuruu) [inline hnet/handler.go:274]: I think the default status should be `NotSubmitted`, if I recall correctly. Unknown was for checksum mismatches.

Comment (Lekuruu) [reply hnet/handler.go:274]: I am not too sure, better if you check first.

Comment (ascenttree) [reply hnet/handler.go:274]: `NeedsUpdate` takes priority over ranked status anyways, so this shouldn't really matter in the long run.

Comment (Lekuruu) [reply hnet/handler.go:274]: It's just how osu! does it, so no clue.

Comment (ascenttree) [reply hnet/handler.go:274]: I think this should be fine for now, if it's not in db, then for all we know it's unsubmitted, and if checksums don't match, we can assume client's map is outdated

**Review by Lekuruu — APPROVED** [review]

**Pull request merged**